### PR TITLE
[WIP] Prune stale environments.

### LINKED ISSUE #1555: Prune unused non-production environments (codename 'Arnie')
## Describe the feature

At the moment there exists in Lagoon an 'auto-idler' to which will spin down non-production pods when there is no traffic to the environment in a given time period (around 4 hours from memory).

This feature is taking this concept 1 step further and removes the environment, reclaiming any PVs and generally freeing up the cluster from the burden of looking after unused environments.

It is not uncommon on some projects to have extremely high non-production environment limits, up to 50 is not uncommon to see in use on a single site. Most of these branches/PR environments just lie around, doing nothing except using up auto-idler time, and PV claims.

The logic _could_ work something like this:

* Is the environment considered **non-production**?
* Has there been **no deployments in the last X days** (e.g. 90 days)
* If yes to both, then remove the environment

On an existing large cluster, I did a Kibana query, and this number represents around 15% of all projects deployed to OpenShift. This would represent a tangible increase in overall cluster resources.

## What can users do to prevent an environment being removed?

There are many ways to avoid the environment being terminated:

* Push a commit to the appropriate branch or PR
* Trigger a redeployment (e.g. through GraphQL, or the Lagoon CLI)
* Perhaps even a button in the Lagoon UI to force the above

Potentially we could also safe guard certain patterns as well, e.g. `/^dev(elop)?/` or `/test/` from the Arnie script.

## Extra for experts

Have the Arnie script configurable per cluster, even to the point of disabling it altogether if needs be. This way each customer with a dedicated cluster can set what is sensible for them. Perhaps 90 days could be the default however.

### PR DESCRIPTION
**Pull request by steveworley**
- Adds a new service that will make a Lagoon API request to remove the environment.

# Closing issues
Closes #1555

### CODE REVIEW

File: services/environment-pruner/prune.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+# set -e -o pipefail
+
+# Create a JWT admin token to talk to the API.
+API_ADMIN_JWT_TOKEN=$(./create_jwt.sh)
+BEARER="Authorization: bearer $API_ADMIN_JWT_TOKEN"
+
+CUTOFF_DATE=$(date +'%Y-%m-%d %H:%M:%S' --date '-90 days')
```
Comment (tobybellwood) [inline services/environment-pruner/prune.sh]: should this `-90` be hardcoded, or definable in a variable?

Comment (steveworley) [reply services/environment-pruner/prune.sh]: Ideally a variable :) the other thing I'm not 100% certain about is if just triggering a delete mutation is enough to remove the environment (it's how we've been doing it) but there might be some oc magic that is needed as well.

File: services/environment-pruner/prune.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+      echo "$OPENSHIFT_URL - $PROJECT_NAME: handling environment $ENVIRONMENT_NAME"
+
+      if [ "$ENVIRONMENT_NAME" = "$PRODUCTION_BRANCH" ]; then
+        # This should only ever remove non-prods.
+        echo "$PROJECT_NAME: $ENVIRONMENT_NAME is the production environment"
+        continue;
+      fi
+
+      Skip if no openshift project exists.
```
Comment (seanhamlin) [inline services/environment-pruner/prune.sh]: should be a comment yeah?

File: services/environment-pruner/prune.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    echo "$project" | jq -c '.environments[]' | while read environment
+    do
+      ENVIRONMENT_OPENSHIFT_PROJECTNAME=$(echo "$environment" | jq -r '.openshiftProjectName')
+      ENVIRONMENT_NAME=$(echo "$environment" | jq -r '.name')
+      UPDATED_DATE=$(echo "$environment" | jq -r '.updated')
+
+      echo "$OPENSHIFT_URL - $PROJECT_NAME: handling environment $ENVIRONMENT_NAME"
+
+      if [ "$ENVIRONMENT_NAME" = "$PRODUCTION_BRANCH" ]; then
```
Comment (seanhamlin) [inline services/environment-pruner/prune.sh]: string comparison in bash is weird. a single equals always makes me twitch.

**Comment by seanhamlin** [comment]
Love this @steveworley ! I salute you.

File: services/environment-pruner/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    && npm config set unsafe-perm true \
+    && npm -g install jwtgen
+
+COPY create_jwt.sh prune.sh /
+
+ENV JWTSECRET=super-secret-string \
+    JWTAUDIENCE=api.dev \
+    PROJECT_REGEX=".+"  \
+    ROUTER_LOG_INTERVAL=4h \
+    POD_RUN_INTERVAL=14400 \
```
Comment (Schnitzel) [inline services/environment-pruner/Dockerfile:16]: this environment variable is never used?

File: services/environment-pruner/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    ELASTICSEARCH_URL="http://logs-db-service:9200"
```
Comment (Schnitzel) [inline services/environment-pruner/Dockerfile:17]: is never used

File: services/environment-pruner/.lagoon.yml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+              memory: 10Mi
+    test: false
+    triggers:
+    - type: ConfigChange
+- apiVersion: batch/v1beta1
+  kind: CronJob
+  metadata:
+    name: cronjob-remote
+  spec:
+    schedule: "*/15 * * * *"
```
Comment (Schnitzel) [inline services/environment-pruner/.lagoon.yml]: I think this is a bit too agressive to run this every 15min?

File: services/environment-pruner/.lagoon.yml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    suspend: true # we're running cronjobs now via the lagoon cronjob system
```
Comment (Schnitzel) [inline services/environment-pruner/.lagoon.yml]: this will actually never run the conjob, not sure this is what we want?

**Review by Schnitzel — CHANGES_REQUESTED** [review]
a nice start, thanks @steveworley for this.

Just a question overall: Can we somehow make sure that this is not enabled by default? Such a service might makes sense for some usecases, but it could also lead into very very confused users of Lagoon.

Also I think we need to create a system to disable this pruning on a project/environment base, like we do for the `autoIdle`, see https://github.com/amazeeio/lagoon/search?q=autoidle&type=Code

File: services/environment-pruner/.lagoon.yml (hunk 6 above)
Comment (seanhamlin) [reply services/environment-pruner/.lagoon.yml]: Agree, when looking at a timespan measured at 90 days, running it once a day seems OK.

**Comment by steveworley** [comment]
Thanks! I just wanted to get a commit up and a starting point validated to make sure it was a direction everyone was happy with :)

> Also I think we need to create a system to disable this pruning on a project/environment base, like we do for the `autoIdle`, see https://github.com/amazeeio/lagoon/search?q=autoidle&type=Code

Agreed - I have some local changes that looks to update the Lagoon schema definition for projects/environments to allow pruning.

File: services/environment-pruner/prune.sh (hunk 2 above)
Comment (steveworley) [reply services/environment-pruner/prune.sh]: Whoops! Yeah - I went a bit crazy with my uncommenting while I was testing.

**Comment by tobybellwood** [comment]
There are two angles for consideration here - the management of environments might be better managed by the Lagoon administrator, and this task shouldn't require a dedicated service. Going to close here whilst we rethink

**Pull request closed without merging**